# 0036-invalid-keyword.patch
#
# DE DONDE SALE. Parche 0030 de los 37 de syntax-transform/3.1/ del PR https://github.com/mkxp-z/mkxp-z/pull/304
# de mkxp-z, de white-axe (rama syntax-transform, 0be32ad119b054749cf3db8c72434f8fa1ac83ea).
# Ensenan al parser y al nucleo de Ruby 3.1 la gramatica y la semantica de
# Ruby 1.8, que es en lo que estan escritos cinco de los ocho juegos del censo.
#
# COMO SE APLICA. Sobre src/ruby-3.1.7 DESPUES de los parches 0001-0005 y en
# orden de numero: el 0006 (prelude) monta el mecanismo que los otros 36 usan,
# y varios de ellos tocan el mismo fichero, asi que el orden no es decorativo.
#
#     cd src/ruby-3.1.7
#     sed -i 's/\r$//' string.c
#     for n in $(seq 6 42); do
#       patch -p1 -i "../../patches/$(printf '%04d' $n)-"*.patch || break
#     done
#     rm -f parse.c id.h insns.inc insns_info.inc opt_sc.inc optinsn.inc \
#           optunifs.inc vm.inc vmtc.inc mjit_compile.inc
#
# Ese `rm` NO es opcional y cuesta una tarde descubrirlo. El tarball de Ruby
# trae parse.c y los .inc YA GENERADOS en el propio directorio de fuentes, y
# un `#include "insns.inc"` desde compile.c mira primero la carpeta del fichero
# que incluye: aunque el build regenere los suyos al lado del Makefile, gana el
# viejo. Con parse.c viejo los ocho parches de gramatica no llegan al binario y
# NADA lo dice; con los .inc viejos, al menos, no compila
# ('YARVINSN_mkxp_syntax_transform_raise_invalid_break' undeclared).
#
# El cuerpo del diff, de aqui abajo, es el del PR VERBATIM: los 37 aplicaron
# sin fuzz y sin rechazos sobre este arbol, asi que no habia nada que adaptar.
# Lo unico que hay que vigilar es string.c, que en este repositorio esta en
# CRLF y al que `patch` le deja el fichero entero en LF; se normaliza antes,
# para que ese cambio sea deliberado y no un efecto secundario.
#
# COMPROBADO. Con los 37 puestos, bison 3.8.2 acepta la gramatica (parse.y
# declara %expect 0: cualquier conflicto nuevo habria sido un error, y no hubo
# ninguno) y CRuby 3.1.7 cruza-compila para aarch64 con devkitA64 16.1.0 hasta
# libruby-static.a (44,7 MB), con los simbolos del mecanismo dentro
# (mkxp_syntax_transform_set_next_eval, mkxp_ec_is_syntax_transform_active,
# mkxp_syntax_transform_target_ruby_version_*) y 14 instrucciones nuevas en
# insns.inc. No se ha ejecutado: eso pide consola.
#
# QUE HACE FALTA ADEMAS. Estos parches por si solos NO cambian el
# comportamiento: mkxp_syntax_transform_target_ruby_version_* nace en -1 y la
# transformacion queda apagada, asi que el interprete se comporta como el de
# hoy. Encenderla es cosa del lado de mkxp-z (armar
# mkxp_syntax_transform_set_next_eval(1) alrededor del eval de cada seccion de
# Scripts.rxdata y fijar la version objetivo desde mkxp.json), que en este
# repositorio todavia no esta hecho.
#
# Licencia: GPL-2.0-or-later, la del PR. La cabecera original va debajo.
# ---------------------------------------------------------------------------

#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, invalid uses of `break`, `next`, `redo`, `retry`, `return` and
# `yield` raise `LocalJumpError`s when the keyword is reached at execution time.
# In later versions of Ruby, these cause syntax errors at compile time instead.
# So this code snippet throws a syntax error in Ruby versions later than 1.8 but
# doesn't in Ruby 1.8.
#
#   def f1
#     break
#   end
#   def f2
#     next
#   end
#   def f3
#     redo
#   end
#   def f4
#     retry
#   end
#   if false
#     class A
#       return
#     end
#   end
#   if false
#     yield
#   end

--- a/compile.c
+++ b/compile.c
@@ -7424,6 +7424,10 @@ compile_break(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, i
             }
             return COMPILE_OK;
 	}
+	if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	    ADD_INSN(ret, line_node, mkxp_syntax_transform_raise_invalid_break);
+	    return COMPILE_OK;
+	}
 	COMPILE_ERROR(ERROR_ARGS "Invalid break");
 	return COMPILE_NG;
     }
@@ -7497,6 +7501,10 @@ compile_next(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, in
 	    }
 	}
 	else {
+	    if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	        ADD_INSN(ret, line_node, mkxp_syntax_transform_raise_invalid_next);
+	        return COMPILE_OK;
+	    }
 	    COMPILE_ERROR(ERROR_ARGS "Invalid next");
 	    return COMPILE_NG;
 	}
@@ -7566,6 +7574,10 @@ compile_redo(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, in
 	    }
 	}
 	else {
+	    if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	        ADD_INSN(ret, line_node, mkxp_syntax_transform_raise_invalid_redo);
+	        return COMPILE_OK;
+	    }
 	    COMPILE_ERROR(ERROR_ARGS "Invalid redo");
 	    return COMPILE_NG;
 	}
@@ -7587,6 +7599,10 @@ compile_retry(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, i
 	}
     }
     else {
+	if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	    ADD_INSN(ret, line_node, mkxp_syntax_transform_raise_invalid_retry);
+	    return COMPILE_OK;
+	}
 	COMPILE_ERROR(ERROR_ARGS "Invalid retry");
 	return COMPILE_NG;
     }
@@ -7766,6 +7782,11 @@ compile_return(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node,
 		type = ISEQ_TYPE_METHOD;
 	    }
 	    break;
+	  case ISEQ_TYPE_CLASS:
+	    if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	        ADD_INSN(ret, line_node, mkxp_syntax_transform_raise_invalid_return);
+	        return COMPILE_OK;
+	    }
 	  default:
 	    break;
 	}
@@ -8873,6 +8894,10 @@ compile_yield(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, i
       case ISEQ_TYPE_TOP:
       case ISEQ_TYPE_MAIN:
       case ISEQ_TYPE_CLASS:
+	if (mkxp_iseq_is_syntax_transform_active(iseq, 1, 8, -1)) {
+	    ADD_INSN(ret, node, mkxp_syntax_transform_raise_invalid_yield);
+	    return COMPILE_OK;
+	}
 	COMPILE_ERROR(ERROR_ARGS "Invalid yield");
 	return COMPILE_NG;
       default: /* valid */;
--- a/insns.def
+++ b/insns.def
@@ -1531,3 +1531,57 @@ answer
 {
     ret = INT2FIX(42);
 }
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_break
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "unexpected break"));
+}
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_next
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "unexpected next"));
+}
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_redo
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "unexpected redo"));
+}
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_retry
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "retry outside of rescue clause"));
+}
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_return
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "unexpected return"));
+}
+
+DEFINE_INSN
+mkxp_syntax_transform_raise_invalid_yield
+()
+()
+()
+{
+    rb_exc_raise(rb_exc_new_cstr(rb_eLocalJumpError, "no block given"));
+}
--- a/parse.y
+++ b/parse.y
@@ -3668,7 +3668,7 @@ k_end		: keyword_end
 
 k_return	: keyword_return
 		    {
-			if (p->ctxt.in_class && !p->ctxt.in_def && !dyna_in_block(p))
+			if (!mkxp_parser_is_syntax_transform_active(p, 1, 8, -1) && p->ctxt.in_class && !p->ctxt.in_def && !dyna_in_block(p))
 			    yyerror1(&@1, "Invalid return in class/module body");
 		    }
 		;
